File: nodes/extract-text-content/extract-text-content.process.js
export default async ({inputs, settings, config}) => {
    try {
        const input = inputs.input;
        if (!input) {
            throw new Error("Input is required for text extraction");
        }

        const includeRole = settings.include_role === true;
        const roleFormat = settings.role_format || "{role}: ";
        const separator = settings.separator || "\n\n";

        // Helper function to extract text from content
        const extractTextFromContent = (content) => {
            if (typeof content === 'string') {
                return content;
            } else if (Array.isArray(content)) {
                let text = '';
                for (const item of content) {
                    if (item && typeof item === 'object' && item.type === 'text' && item.text) {
                        text += item.text;
                    }
                }
                return text;
            } else {
                return String(content || '');
            }
        };

        // Helper function to format role prefix
        const formatRolePrefix = (role) => {
            if (!includeRole || !role) return '';
            return roleFormat.replace('{role}', role.toUpperCase());
        };

        let extractedTexts = [];
        let messageCount = 0;

        if (typeof input === 'string') {
            // Direct string input
            extractedTexts.push(input);
            messageCount = 1;
        } else if (Array.isArray(input)) {
            // Array of messages
            for (const message of input) {
                if (typeof message === 'string') {
                    // String message - no role assumption
                    extractedTexts.push(message);
                    messageCount++;
                } else if (message && typeof message === 'object' && message.content !== undefined) {
                    // Message object
                    const text = extractTextFromContent(message.content);
                    if (text.trim()) {
                        const rolePrefix = formatRolePrefix(message.role);
                        extractedTexts.push(rolePrefix + text);
                        messageCount++;
                    }
                }
            }
        } else if (input && typeof input === 'object' && input.content !== undefined) {
            // Single message object
            const text = extractTextFromContent(input.content);
            if (text.trim()) {
                const rolePrefix = formatRolePrefix(input.role);
                extractedTexts.push(rolePrefix + text);
                messageCount = 1;
            }
        } else {
            // Fallback for other types
            const text = String(input);
            extractedTexts.push(text);
            messageCount = 1;
        }

        const result = extractedTexts.join(separator);

        return {
            text: result,
            message_count: messageCount
        };

    } catch (error) {
        console.log('Extract Text Content error:', error);
        throw new Error(`Extract Text Content error: ${error.message}`);
    }
};

File: nodes/extract-text-content/extract-text-content.process.py
from typing import Any, Dict, List, Union

async def process(inputs: Dict[str, Any], settings: Dict[str, Any], config: Dict[str, Any], nodeConfig: Dict[str, Any]) -> Dict[str, Any]:
    try:
        input_data = inputs.get('input')
        if not input_data:
            raise Exception("Input is required for text extraction")

        include_role = settings.get('include_role', False)
        role_format = settings.get('role_format', '{role}: ')
        separator = settings.get('separator', '\n\n')

        # Helper function to extract text from content
        def extract_text_from_content(content):
            if isinstance(content, str):
                return content
            elif isinstance(content, list):
                text = ''
                for item in content:
                    if isinstance(item, dict) and item.get('type') == 'text' and 'text' in item:
                        text += item['text']
                return text
            else:
                return str(content or '')

        # Helper function to format role prefix
        def format_role_prefix(role):
            if not include_role or not role:
                return ''
            return role_format.replace('{role}', role.upper())

        extracted_texts = []
        message_count = 0

        if isinstance(input_data, str):
            # Direct string input
            extracted_texts.append(input_data)
            message_count = 1
        elif isinstance(input_data, list):
            # Array of messages
            for message in input_data:
                if isinstance(message, str):
                    # String message - no role assumption
                    extracted_texts.append(message)
                    message_count += 1
                elif isinstance(message, dict) and 'content' in message:
                    # Message object
                    text = extract_text_from_content(message['content'])
                    if text.strip():
                        role_prefix = format_role_prefix(message.get('role'))
                        extracted_texts.append(role_prefix + text)
                        message_count += 1
        elif isinstance(input_data, dict) and 'content' in input_data:
            # Single message object
            text = extract_text_from_content(input_data['content'])
            if text.strip():
                role_prefix = format_role_prefix(input_data.get('role'))
                extracted_texts.append(role_prefix + text)
                message_count = 1
        else:
            # Fallback for other types
            text = str(input_data)
            extracted_texts.append(text)
            message_count = 1

        result = separator.join(extracted_texts)

        return {
            "text": result,
            "message_count": message_count
        }

    except Exception as error:
        print(f'Extract Text Content error: {error}')
        raise Exception(f"Extract Text Content error: {str(error)}")
